### PR TITLE
feat: Add `Builder::permissions()` method.

### PR DESCRIPTION
**Pull request by Byron**
This PR allows the user to control the file mode set for newly created `NameTempfile` instances.

This is paramount for when tempfiles are used as scratch space that will be persisted and renamed into the final destination once all writes are completed. This is an effective way to prevent half-written files on disk and to assure that readers can only observe the final result of a change.

This paradigm is implemented in the `gix-lock` crate which is used by the `gix-ref` crate for editing Git references. Currently these are created with `0o600` permissions, which can cause [issues when using `sudo`](https://github.com/stacked-git/stgit/issues/413) in downstream applications.

When this PR is merged, `gix-ref` can be adjusted to allow setting the mode [just like Git does](https://github.com/git/git/blob/master/lockfile.h#L175-L181).

### Review Notes

I did my best to adapt the least amount of code while staying true to the current organisation.  Initially I named the new method `mode`, but realized that `std::fs::Permissions` exists which would allow the API to be more general, despite only one available platform for implementation.

Let me explain the commits in some detail and why it's one more than needed:

* the first commit is a failing test that adds the new `Builder` method without an implementation. CI is red to show it's not working.
* the second commits implements it to be least invasive, which violates the principle that platform dependent code should remain in `imp`. CI is green.
* the third commit implements it so that `permissions` can be passed down to the platforms, which is more invasive, yet cleaner. CI is also green.

Please let me know which way you prefer as its also a trade-off.

### CODE REVIEW

**Comment by NobodyXu** [comment]
@Byron There's a better way to create tempfile that persists after writing is done on Linux using `O_TMPFILE`.

While it only supports ext2, ext3, ext4, UDF, Minix, tmpfs 
XFS (Linux 3.15), Btrfs (Linux 3.16), F2FS (Linux 3.16), ubifs (Linux 4.9), and needs either `CAP_DAC_READ_SEARCH` or a procfs to be mounted, it's still a pretty good solution.

It prevents any other process from manipulating the file (except through procfs) until the caller process marked it as ready, and it removed the file on process exit even if the rust destructor is not called (e.g. panic during unwinding, calling `exit` directly)

```
O_TMPFILE (since Linux 3.11)
              Create an unnamed temporary regular file.  The pathname
              argument specifies a directory; an unnamed inode will be
              created in that directory's filesystem.  Anything written
              to the resulting file will be lost when the last file
              descriptor is closed, unless the file is given a name.

              O_TMPFILE must be specified with one of O_RDWR or O_WRONLY
              and, optionally, O_EXCL.  If O_EXCL is not specified, then
              linkat(2) can be used to link the temporary file into the
              filesystem, making it permanent, using code like the
              following:

                  char path[PATH_MAX];
                  fd = open("/path/to/dir", O_TMPFILE | O_RDWR,
                                          S_IRUSR | S_IWUSR);

                  /* File I/O on 'fd'... */

                  linkat(fd, "", AT_FDCWD, "/path/for/file", AT_EMPTY_PATH);

                  /* If the caller doesn't have the CAP_DAC_READ_SEARCH
                     capability (needed to use AT_EMPTY_PATH with linkat(2)),
                     and there is a proc(5) filesystem mounted, then the
                     linkat(2) call above can be replaced with:

                  snprintf(path, PATH_MAX,  "/proc/self/fd/%d", fd);
                  linkat(AT_FDCWD, path, AT_FDCWD, "/path/for/file",
                                          AT_SYMLINK_FOLLOW);
                  */

              In this case, the open() mode argument determines the file
              permission mode, as with O_CREAT.

              Specifying O_EXCL in conjunction with O_TMPFILE prevents a
              temporary file from being linked into the filesystem in
              the above manner.  (Note that the meaning of O_EXCL in
              this case is different from the meaning of O_EXCL
              otherwise.)

              There are two main use cases for O_TMPFILE:

              •  Improved tmpfile(3) functionality: race-free creation
                 of temporary files that (1) are automatically deleted
                 when closed; (2) can never be reached via any pathname;
                 (3) are not subject to symlink attacks; and (4) do not
                 require the caller to devise unique names.

              •  Creating a file that is initially invisible, which is
                 then populated with data and adjusted to have
                 appropriate filesystem attributes (fchown(2),
                 fchmod(2), fsetxattr(2), etc.)  before being atomically
                 linked into the filesystem in a fully formed state
                 (using linkat(2) as described above).

              O_TMPFILE requires support by the underlying filesystem;
              only a subset of Linux filesystems provide that support.
              In the initial implementation, support was provided in the
              ext2, ext3, ext4, UDF, Minix, and tmpfs filesystems.
              Support for other filesystems has subsequently been added
              as follows: XFS (Linux 3.15); Btrfs (Linux 3.16); F2FS
              (Linux 3.16); and ubifs (Linux 4.9)
```

**Comment by Stebalien** [comment]
@NobodyXu I agree if you're only target is Linux, but it's still useful to have a platform-independent way to do this (to, e.g., support MacOS which hasn't seen any filesystem API improvements in, basically ever).

File: src/lib.rs
Comment (Stebalien) [inline src/lib.rs:428]: We need to implement this before we merge this patch, unfortunately. This builder is used to create temporary directories as well, and I don't really trust users to read the docs.

File: src/lib.rs
Comment (Stebalien) [inline src/lib.rs:423]: This is annoying... but there's not much we can do here. Well, maybe check `.readonly()` on windows and fail if it's set?

File: src/lib.rs
Comment (Stebalien) [reply src/lib.rs:428]: See the discussion in #61.

**Comment by Stebalien** [comment]
Thanks! And yes, I prefer the final version (more invasive approach).

File: src/lib.rs
Comment (Byron) [reply src/lib.rs:428]: I took a look at this and saw that `std::fs::create_dir()` is used to create the directory, I don't think the standard library has a way to handle this.
Did you have an idea on how to do this? Maybe a `rustix` `chmod` call?

File: src/lib.rs
Comment (Byron) [reply src/lib.rs]: That's a great idea! I have added this check in the most recent commit.

**Comment by Byron** [comment]
Thanks for the quick review! If CI passes I have addressed the windows portion of it, but will definitely need some guidance to deal with the directory. If you have ideas and want to tinker, please feel free to push into this branch - sometimes that's easiest.

@NobodyXu Thanks for sharing! It's definitely something to consider for the time when performance needs to be optimised further. Personally I am most interested in doing what Git does at first, which seems to be pretty much what's happening here.

File: src/lib.rs
Comment (Stebalien) [reply src/lib.rs:429]: I think we'll need to use `rustix::fs::mkdir`.

**Comment by Byron** [comment]
I have added mode-support for directories thanks to your hint with `rustix::fs::mkdir` and it appears to work. Further, trying to change permissions on unsupported platforms will result in an error that as well.
When CI is green it's ready for review, with the remark that I kept `dir.rs` nimble, with a structure sufficient for the task even though it's not quite the same as in `file`.

File: src/dir.rs
Comment (Stebalien) [inline src/dir.rs:492]: I believe the correct thing to pass here is 0777, and let the umask deal with it (we're being inconsistent with tempfile.... but changing that now isn't a great idea).

File: src/dir.rs
Comment (Stebalien) [inline src/dir.rs:507]: You can write this as:

```rust
                .map_err(Into::into)
                .with_err_path(|| &path)
```

Then revert the change to make the fields public.

File: src/dir.rs
Comment (Stebalien) [inline src/dir.rs:478]: I'd prefer it if we created the entire directory structure.

**Comment by Stebalien** [comment]
So... somehow I missed this but: https://doc.rust-lang.org/std/fs/struct.DirBuilder.html#method.mode. We don't need to use rustix here.

File: src/dir.rs
Comment (Byron) [reply src/dir.rs:478]: I can do that, even though it's a change that might have performance implications. For that reason alone I wouldn't do it without also introducing a new method for it, and do so in a separate PR.

File: src/dir.rs
Comment (Byron) [reply src/dir.rs:492]: With `std::fs::DirBuilder` we have the luxury of not specifying it unless it's overridden, I will do that.

**Comment by Byron** [comment]
> So... somehow I missed this but: https://doc.rust-lang.org/std/fs/struct.DirBuilder.html#method.mode. We don't need to use rustix here.

Oh my, I missed that too! I clearly was biased when researching this topic and "didn't believe in mode being available for directories" 🤦‍♂️. Lesson learned, and the code was adjusted to something much better.

I made a note that recommends to not recursively create the directory by default but leave that to a separate PR along with a new flag in the builder maybe to control this, to keep the previous behaviour exactly the same. Right now, this still is the case.

If CI is green, this PR should be ready for re-review.

File: src/dir.rs
Comment (Stebalien) [reply src/dir.rs:478]: Wait, no. I mean: put the modules in different files/directories (`imp/...`). Sorry, I realize that that was confusing.

Comment (Byron) [reply src/dir.rs]: Ah, right, I could have guessed it. And it's done now. A note regarding naming: It feels like `any` should be `other`, but I decided not to do that as `other` means unsupported in the `file` folder, while here it is a catch-all.

File: src/error.rs
Comment (Stebalien) [inline src/error.rs:7]: This change can be revered.

File: src/dir/imp/unix.rs
Comment (Stebalien) [inline src/dir/imp/unix.rs:17]:
```suggestion
        .with_err_path(|| &path)
```

(the path is cloned automatically if there's an error)

File: src/lib.rs
Comment (Stebalien) [reply src/lib.rs:430]: This comment is now out of date.

File: src/dir.rs
Comment (Stebalien) [reply src/dir.rs]: Hm. Yeah, I may go through and do a rename later but those are implementation details so.. meh.

**Comment by Stebalien** [comment]
And with those changes, this should be good to go. Thanks!

**Comment by Byron** [comment]
Thanks a lot for the QA and sorry for the sloppiness, at least 2 of 3 I should have caught.
But here we are, and I think that's the version that is ready to go.

**Review by Stebalien — APPROVED** [review]
❤️

**Comment by Stebalien** [comment]
> Thanks a lot for the QA and sorry for the sloppiness, at least 2 of 3 I should have caught.

Don't be. That'll just make me feel worse about my PRs and that's why we have review 😆.

> PS: A new release with these changes would help, as that way everything can trickle downstream to StackedGit. Thanks again.

Will do.

**Pull request merged**
